messages2/src/a2a.rs

```diff
@@ -0,0 +1,190 @@
+        S: Serializer,
+        S::Error: From<M::Error>,
+    {
+        match self {
+            Self::Routing(v) => v.delayed_serialize(state, closure),
+            Self::Connection(v) => v.delayed_serialize(state, closure),
+            Self::Revocation(v) => v.delayed_serialize(state, closure),
+            Self::CredentialIssuance(v) => v.delayed_serialize(state, closure),
+            Self::ReportProblem(v) => v.delayed_serialize(state, closure),
+            Self::PresentProof(v) => v.delayed_serialize(state, closure),
+            Self::TrustPing(v) => v.delayed_serialize(state, closure),
+            Self::DiscoverFeatures(v) => v.delayed_serialize(state, closure),
+            Self::BasicMessage(v) => v.delayed_serialize(state, closure),
+            Self::OutOfBand(v) => v.delayed_serialize(state, closure),
+        }
+    }
+}
+
+/// Custom [`Deserialize`] impl for [`A2AMessage`] to use the `@type` as internal tag,
+/// but deserialize it to a [`MessageType`].
+/// 
+/// For readability, the [`MessageType`] matching is done in the [`DelayedSerde::delayed_deserialize`] method.
+//
+// Yes, we're using some private serde constructs. Here's why I think this is okay:
+//  1) This emulates the derived implementation with the #[serde(tag = "@type")] attribute,
+//     but uses [`MessageType`] instead of some [`Field`] struct that serde generates.
+//
+//  2) Without this, the implementation would either rely on something inefficient such as [`Value`] as an intermediary,
+//     use some custom map which fails on duplicate entries as intermediary or basically use [`serde_value`]
+//     which seems to be an old replica of [`Content`] and [`ContentDeserializer`] and require a pretty much
+//     copy paste of [`TaggedContentVisitor`].
+//
+//  3) Exposing these parts as public is in progress from serde. When that will happen is still unknown.
+//     See: https://github.com/serde-rs/serde/issues/741
+//     With [`serde_value`] lacking activity and not seeming to get integrated into [`serde`], this will most likely resurface.
+//
+//  4) Reimplementing this on breaking semver changes is as easy as expanding the derived [`Deserialize`] impl and altering it a bit.
+//     And if that fails, the 2nd argument will still be viable.
+//
+//
+// In the event of a `serde` version bump and this breaking, the fix is a matter of
+// implementing a struct such as:
+// ```
+// #[derive(Deserialize)]
+// #[serde(tag = "@type")]
+// enum MyStruct {
+//     Var(u8),
+//     Var2(u8)    
+// }
+// ```
+//
+// Then analyze the expanded [`Deserialize`] impl and adapt the actual implementation below.
+impl<'de> Deserialize<'de> for A2AMessage {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        use serde::__private::de::{ContentDeserializer, TaggedContentVisitor};
+
+        let tag_visitor = TaggedContentVisitor::<MessageType>::new("@type", "internally tagged enum A2AMessage");
```

Comment:
perhaps you could explain not only why it's okay to use `TaggedContentVisitor`, but what it's doing - it's not easily lookup-able since it's not yet public API. Same would go for `ContentDeserializer`

Reply:
Done.